# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/GamepadPlatformService.cpp
# Commit: 4d4247338bf6
# Full Hash: 4d4247338bf6ebab2ebc69a89b7a4cb610cc5fee
# Author: Randell Jesup <rjesup@mozilla.com>
# Date: 2024-12-10 16:36:00
# Regressor Bug: 1657404
# File Overlap Count: 1
# Description:
#   Bug 1657404: Back out changeset 349854c5ad56 (gamepad) r=cmartin
#   
#   This changeset was left in when the rest of bug 1657404 was backed out in
#   2020, and causes at least some Windows machines to not see gamepads
#   
# ==============================================================================

diff -r ade719ece0a8 -r 4d4247338bf6 dom/gamepad/GamepadPlatformService.cpp
--- a/dom/gamepad/GamepadPlatformService.cpp	Tue Dec 10 02:32:07 2024 +0000
+++ b/dom/gamepad/GamepadPlatformService.cpp	Tue Dec 10 02:56:44 2024 +0000
@@ -28,44 +28,33 @@
 
 }  // namespace
 
-// static
-GamepadMonitoringState& GamepadMonitoringState::GetSingleton() {
-  static GamepadMonitoringState sInstance{};
-  return sInstance;
+GamepadPlatformService::MonitoringState::~MonitoringState() {
+  AssertIsOnBackgroundThread();
+  MOZ_RELEASE_ASSERT(mObservers.IsEmpty());
 }
 
-void GamepadMonitoringState::AddObserver(GamepadTestChannelParent* aParent) {
-  AssertIsOnBackgroundThread();
-  MOZ_ASSERT(aParent);
-  MOZ_ALWAYS_TRUE(mObservers.append(aParent));
-}
-
-void GamepadMonitoringState::RemoveObserver(GamepadTestChannelParent* aParent) {
+void GamepadPlatformService::MonitoringState::AddObserver(
+    WeakPtr<GamepadTestChannelParent> aParent) {
   AssertIsOnBackgroundThread();
   MOZ_ASSERT(aParent);
-
-  WeakPtr<GamepadTestChannelParent>* observer = nullptr;
-
-  for (auto& item : mObservers) {
-    if (item == aParent) {
-      observer = &item;
-    }
-  }
-
-  MOZ_ASSERT(
-      observer,
-      "Attempted to remove a GamepadTestChannelParent that was never added");
-
-  std::swap(*observer, mObservers.back());
-  mObservers.popBack();
+  MOZ_ASSERT(!mObservers.Contains(aParent));
+  mObservers.AppendElement(std::move(aParent));
 }
 
-bool GamepadMonitoringState::IsMonitoring() const {
+void GamepadPlatformService::MonitoringState::RemoveObserver(
+    GamepadTestChannelParent* aParent) {
+  AssertIsOnBackgroundThread();
+  MOZ_ASSERT(aParent);
+  MOZ_ASSERT(mObservers.Contains(aParent));
+  mObservers.RemoveElement(aParent);
+}
+
+bool GamepadPlatformService::MonitoringState::IsMonitoring() const {
   AssertIsOnBackgroundThread();
   return mIsMonitoring;
 }
 
-void GamepadMonitoringState::Set(bool aIsMonitoring) {
+void GamepadPlatformService::MonitoringState::Set(bool aIsMonitoring) {
   AssertIsOnBackgroundThread();
 
   if (mIsMonitoring != aIsMonitoring) {
@@ -273,7 +262,7 @@
 
   StartGamepadMonitoring();
 
-  GamepadMonitoringState::GetSingleton().Set(true);
+  mMonitoringState.Set(true);
 }
 
 void GamepadPlatformService::RemoveChannelParent(
@@ -293,7 +282,7 @@
     }
   }
 
-  GamepadMonitoringState::GetSingleton().Set(false);
+  mMonitoringState.Set(false);
 
   StopGamepadMonitoring();
   ResetGamepadIndexes();